### PR TITLE
kv: fix introduced TestMultiRangeBoundedWithCompletedUnboundedScan

### PR DESCRIPTION
**Pull request by vivekmenezes**
I added this test that failed, and found a fix that I'm not a 100% sure is the right solution. See the new comment in the code.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7720)

### CODE REVIEW

**Comment by tschottdorf** [comment]
Good catch, though the suggested solution is wrong (it fixes this test, and it's a bit worrying that it doesn't break any of the other tests, but maybe you could find one that will break? The breakage should be that you're just stopping with the batch once a request has saturated).

I went ahead and looked at the real solution. You may want to replace this branch with https://github.com/cockroachdb/cockroach/tree/tschottdorf/batch-saturation (the diff is quite small).

I'm also sure I had to duplicate some work you had already done in researching the root cause. You could have saved me some time with a more accurate description of the test failure prior to your attempted fix (no biggie - just for next time). I did some printf debugging until I found out that we got stuck in an infinite range descriptor lookup loop (which always failed because we attempted to look up `[KeyMax, someKey)`), at which point I looked at `rs.Key` being updated at the end of the inner loop, close to where most of the diff lives.

Maybe you could add some more of these batch saturation cases to your test?

**Comment by vivekmenezes** [comment]
Thanks for fixing this for me Tobi! I was very confused by this code, and I think that this commit is in a much better state with some better comments and a test that's more specific for the case that needed testing. It's mostly copied from your branch. Thanks

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7720#-:-KMPfYLoJox6bOEjPNHJ:1722202432)*

File: kv/dist_sender.go [hunk 1]
```diff
@@ -867,7 +865,10 @@ func (ds *DistSender) sendChunk(ctx context.Context, ba roachpb.BatchRequest) (*
 				}
 				boundedArg, ok := args.(roachpb.Bounded)
 				if !ok {
-					// Non-bounded request. We will have to query all ranges.
+					// Non-bounded request. We will have to continue querying
+					// until this request is satisfied. This request might
+					// have already been saturated at this stage but we have
```
Comment (tschottdorf) [inline kv/dist_sender.go]: I would reserve "saturated" to mean "reached the number of requested entries". You're using it for "the key range has been completely queried".

File: kv/dist_sender_server_test.go [hunk 2]
```diff
@@ -152,6 +155,42 @@ func TestMultiRangeBatchBoundedScans(t *testing.T) {
 	})
 }
 
+// TestMultiRangeBoundedWithSaturatedUnboundedScan runs a batch request with a
+// bounded and unbounded scan, such that the bounded scan gets saturated when
+// the unbounded scan has already been saturated. Additionally, the bound for
```
Comment (tschottdorf) [inline kv/dist_sender_server_test.go]: ditto

**Comment by tschottdorf** [comment]
You're welcome. BTW, my suggestion involving `git rebase --hard` had the
background of allowing me easier review of the current diff and also
preserving authorships accordingly (which matters for git blame).
-- 

-- Tobias

**Comment by vivekmenezes** [comment]
ah ha thing to remember and be remembered

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender.go, line 870 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7720#-KMHy3_Te2yI0cMB38St-r2-870:-KMPpt-AH094McmQ3VjL:54506078) ([raw file](https://github.com/cockroachdb/cockroach/blob/33c5007c6514b7d9f6cb23d5cb43ffdaaec172c2/kv/dist_sender.go#L870)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I would reserve "saturated" to mean "reached the number of requested entries". You're using it for "the key range has been completely queried".
</details>
s/saturated/completed where needed

---

*[kv/dist_sender_server_test.go, line 160 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7720#-KMHy3_Te2yI0cMB38Su-r2-160:-KMPpxCVORgUIumwigh8:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/33c5007c6514b7d9f6cb23d5cb43ffdaaec172c2/kv/dist_sender_server_test.go#L160)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ditto
</details>
fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7720#-:-KMPqgWbgwDSedTm8XF0:951692973)*

**Comment by bdarnell** [comment]
LGTM, but the commit message could use some more detail about the bug being fixed.

**Comment by vivekmenezes** [comment]
Done

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7720#-:-KMQYR59kJB8G-dXDtB0:676635175)*

**Pull request merged**